distfunc.c

```diff
@@ -0,0 +1,50 @@
+#include "postgres.h"
+#include "embedding.h"
+#include "math.h"
+
+dist_t l2_dist_impl(coord_t const* ax, coord_t const* bx, size_t dim)
+{
+	dist_t 		distance = 0.0;
+	for (size_t i = 0; i < dim; i++)
+	{
+		dist_t diff = ax[i] - bx[i];
+		distance += diff * diff;
+	}
+	return distance;
+}
+
+dist_t cosine_dist_impl(coord_t const* ax, coord_t const* bx, size_t dim)
+{
+	dist_t 		distance = 0.0;
+	dist_t 		norma = 0.0;
+	dist_t 		normb = 0.0;
+	for (size_t i = 0; i < dim; i++)
+	{
+		distance += ax[i] * bx[i];
+		norma += ax[i] * ax[i];
+		normb += bx[i] * bx[i];
+	}
+	return 1 - (distance / sqrt(norma * normb));
+}
+
+dist_t manhattan_dist_impl(coord_t const* ax, coord_t const* bx, size_t dim)
+{
+	dist_t 		distance = 0.0;
+	for (size_t i = 0; i < dim; i++)
+	{
+		distance += fabs(ax[i] - bx[i]);
+	}
+	return distance;
+}
+
+static dist_t (*const dist_func_table[])(coord_t const* ax, coord_t const* bx, size_t size) =
+{
+	l2_dist_impl,
+	cosine_dist_impl,
+	manhattan_dist_impl
+};
+
+dist_t hnsw_dist_func(dist_func_t func, coord_t const* ax, coord_t const* bx, size_t dim)
+{
+	return dist_func_table[func](ax, bx, dim);
```

Comment:
You can add `avx2` versions of functions to the tables, and in here have `static bool use_avx2 = __builtin_cpu_supports("avx2");` and then use this flag to offset into table to return avx2 function

Reply:
gcc/clang use AVX2 instructions with -Ofast option. There is still some difference in performance between code generated by GCC and hand written version n using intrinsics: at my `l2_dist` benchmark it is 63 vs 100 nsec, I think mostly because of loop unrolling. I tried to force gcc to perform loop unrolling  optimisation as well, but it doesn't  improve speed.
In any case, as you have noticed:  version with intrinsics is not correctly handling case when dims % 16 != 0.
Certainly it is not so difficult to fix it and it should not affect speed of the case when it is aligned.
But still...

Reply:
GCC/Clang are only allowed to use AVX2 when you specify `-mavx2` in compiler flags, which we don't want. We need to dynamically dispatch on AVX2 or non-AVX2 version.

Reply:
My assumption is that all Intel CPUs used in EC2 support AXV2. So using -mavx2 or even -march=native is acceptable for us.

Reply:
T2 instances for example do not have AVX2, only AVX. But also, `-mavx2` is not valid on ARMs and leads to a lot of compiler warnings. Lastly, I think this extension is not Neon-specific, but a general purpose Postgres extension, and so it should be supported even on peoples' 15-year old laptops.

Reply:
Agree with your arguments.
Changed